tokio-trace/src/field.rs

```diff
@@ -0,0 +1,225 @@
+pub use tokio_trace_core::field::*;
+
+use std::fmt;
+use Metadata;
+
+/// Trait implemented to allow a type to be used as a field key.
+///
+/// **Note**: Although this is implemented for both the [`Field`] type *and* any
+/// type that can be borrowed as an `&str`, only `Field` allows _O_(1) access.
+/// Indexing a field with a string results in an iterative search that performs
+/// string comparisons. Thus, if possible, once the key for a field is known, it
+/// should be used whenever possible.
+pub trait AsField {
+    /// Attempts to convert `&self` into a `Field` with the specified `metadata`.
+    ///
+    /// If `metadata` defines this field, then the field is returned. Otherwise,
+    /// this returns `None`.
+    fn as_field(&self, metadata: &Metadata) -> Option<Field>;
+}
+
+pub trait Record {
+    /// Record a signed 64-bit integer value.
+    fn record_i64<Q: ?Sized>(&mut self, field: &Q, value: i64)
+    where
+        Q: AsField;
+
+    /// Record an umsigned 64-bit integer value.
+    fn record_u64<Q: ?Sized>(&mut self, field: &Q, value: u64)
+    where
+        Q: AsField;
+
+    /// Record a boolean value.
+    fn record_bool<Q: ?Sized>(&mut self, field: &Q, value: bool)
+    where
+        Q: AsField;
+
+    /// Record a string value.
+    fn record_str<Q: ?Sized>(&mut self, field: &Q, value: &str)
+    where
+        Q: AsField;
+
+    /// Record a value implementing `fmt::Debug`.
+    fn record_debug<Q: ?Sized>(&mut self, field: &Q, value: &fmt::Debug)
+    where
+        Q: AsField;
+}
+
+/// A field value of an erased type.
+///
+/// Implementors of `Value` may call the appropriate typed recording methods on
+/// the `Subscriber` passed to `record` in order to indicate how their data
+/// should be recorded.
+pub trait Value {
+    /// Records this value with the given `Subscriber`.
+    fn record<Q: ?Sized, R>(&self, key: &Q, recorder: &mut R)
+    where
+        Q: AsField,
+        R: Record;
+}
+
+/// A `Value` which serializes as a string using `fmt::Display`.
+#[derive(Clone)]
+pub struct DisplayValue<T: fmt::Display>(T);
+
+/// A `Value` which serializes as a string using `fmt::Debug`.
+#[derive(Clone)]
+pub struct DebugValue<T: fmt::Debug>(T);
+
+/// Wraps a type implementing `fmt::Display` as a `Value` that can be
+/// recorded using its `Display` implementation.
+pub fn display<'a, T>(t: T) -> DisplayValue<T>
```

Comment:
Does this `'a` lifetime do anything? Doesn't appear to...

Reply:
Whoops, that's left over from earlier code; removed it in e16bfd0.